src/kernel.rs

```diff
@@ -0,0 +1,98 @@
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::fmt::Result as FmtResult;
+use std::io::Error;
+use std::io::ErrorKind;
+use std::io::Result;
+use std::path::Path;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+use crate::elf::ElfResolver;
+use crate::ksym::KSymResolver;
+use crate::Addr;
+use crate::AddressLineInfo;
+use crate::CacheHolder;
+use crate::FindAddrOpts;
+use crate::SymResolver;
+use crate::SymbolInfo;
+
+
+pub(crate) struct KernelResolver {
+    pub ksymresolver: Option<Rc<KSymResolver>>,
+    pub kernelresolver: Option<ElfResolver>,
+    kallsyms: PathBuf,
+    kernel_image: PathBuf,
+}
+
+impl KernelResolver {
+    pub fn new(
+        kallsyms: &Path,
+        kernel_image: &Path,
+        cache_holder: &CacheHolder,
+    ) -> Result<KernelResolver> {
+        let ksymresolver = cache_holder.get_ksym_cache().get_resolver(kallsyms);
+        let kernelresolver = ElfResolver::new(kernel_image, 0, cache_holder);
+
+        if ksymresolver.is_err() && kernelresolver.is_err() {
+            return Err(Error::new(
+                ErrorKind::NotFound,
+                format!(
+                    "can not load {} and {}",
+                    kallsyms.display(),
+                    kernel_image.display()
+                ),
+            ))
+        }
+
+        Ok(KernelResolver {
+            ksymresolver: ksymresolver.ok(),
+            kernelresolver: kernelresolver.ok(),
+            kallsyms: kallsyms.to_path_buf(),
+            kernel_image: kernel_image.to_path_buf(),
+        })
+    }
+}
+
+impl SymResolver for KernelResolver {
+    fn get_address_range(&self) -> (Addr, Addr) {
+        (0xffffffff80000000, 0xffffffffffffffff)
+    }
```

Comment:
do we use this anywhere? I think these ranges are architecture dependent, but I'm also wondering why we need this, as user is supposed to specify which type of the address (kernel or user) is being symbolized

Seems like some unnecessary assumptions that we might have in the existing code.

Reply:
Yes, it's used to build up the "resolver map" and then later to find the matching resolver. I agree we shouldn't be hard coding these ranges, as this will break on anything not exactly having this layout (including anything not a 64 bit kernel).